File: Data Structures/Array/07 Next Permutation.cpp
//NEXT PERMUTATION (NEXT LEXICOGRAPHICALLY GREATER PERMUTATION)

void nextPermutation(vector<int>& arr) {
	int arr_size =  arr.size();
	int numberBeforePeakFromRight = -1;
	for(int i=arr_size-2; i>=0; i--)
		if(arr[i] < arr[i+1]){
			numberBeforePeakFromRight = i;
			break;
		}      
	if(numberBeforePeakFromRight<0)
		reverse(arr.begin(), arr.end());
	else{
		int smallestNumberGreaterThanPeaksLeft;
		for(int i=arr_size-1; i>numberBeforePeakFromRight; i--)
			if(arr[i] > arr[numberBeforePeakFromRight]){
				smallestNumberGreaterThanPeaksLeft = i;
				break;
			}
		swap(arr[numberBeforePeakFromRight], arr[smallestNumberGreaterThanPeaksLeft]);
		reverse(arr.begin() + numberBeforePeakFromRight + 1, arr.end());
	}
}

//Time Complexity: O(n)
//Space Complexity: O(1)

//Problem Link: https://leetcode.com/problems/next-permutation/

File: Data Structures/Array/07 Next Permutation.java
class Solution {
    public void swap(int arr[], int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
    public void reverse(int arr[], int i, int j){
        while(i < j)    swap(arr, i++, j--);
    }
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int numLeftOfPeak = n-2, i = n-2;
        while(i >= 0 && nums[i] >= nums[i+1]) i--;
        numLeftOfPeak = i;
        if(numLeftOfPeak >= 0){
            int j = n-1;
            while(nums[j] <= nums[i]) j--;
            int numJustGreaterThanPeaksLeft = j;
            swap(nums, numLeftOfPeak, numJustGreaterThanPeaksLeft);
        }
        reverse(nums, numLeftOfPeak + 1, n-1);
    }
}
